tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_cluster_connection_pool.hpp

```diff
@@ -0,0 +1,757 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <inttypes.h>
+#include <unistd.h>
+
+#include <chrono>
+#include <iostream>
+
+#include "tensorflow/core/framework/tensor.h"
+#include <sw/redis++/connection.h>
+#include <sw/redis++/connection_pool.h>
+#include <sw/redis++/redis++.h>
+#include <sw/redis++/sentinel.h>
+
+#include "redis_connection_util.hpp"
+
+using sw::redis::ConnectionOptions;
+using sw::redis::ConnectionPoolOptions;
+using sw::redis::Redis;
+using sw::redis::RedisCluster;
+
+namespace tensorflow {
+namespace recommenders_addons {
+namespace redis_connection {
+template <typename RedisInstance, typename K, typename V>
+class RedisWrapper<RedisInstance, K, V,
+                   typename std::enable_if<
+                       std::is_same<RedisInstance, RedisCluster>::value>::type>
+    : public RedisVirtualWrapper {
+private:
+  ConnectionOptions conn_opts;
+  ConnectionPoolOptions pool_opts;
+
+public:
+  std::shared_ptr<RedisInstance> redis_conn; // for the hungry singleton mode
+
+public:
+  RedisWrapper(RedisInstance &&) = delete;
+  RedisWrapper(const RedisInstance &) = delete;
+  RedisWrapper &operator=(const RedisInstance &) = delete;
+
+  RedisWrapper() // In singleton mode, classes should not be initialized
+                 // through constructor
+  {
+    std::cout << "RedisCluster connection pool constructor called!"
+              << std::endl;
+  }
+
+  ~RedisWrapper() {
+    if (redis_conn == nullptr) {
+      return;
+    }
+    redis_conn.reset();
+    std::cout << "RedisCluster connection pool destructor called!" << std::endl;
+  }
+
+public:
+  std::shared_ptr<RedisInstance> start_conn() {
+    conn_opts.host = redis_connection_params.redis_host_ip;
```

Comment:
If there are multiple master ips and only one ip is configured, is high reliability not guaranteed?